groovy/groovy.editor/src/org/netbeans/modules/groovy/editor/compiler/PerfData.java

```diff
@@ -0,0 +1,226 @@
+    private int fileCount;
+    
+    private PerfData collector;
+    
+    private static ThreadLocal<PerfData> threadInstance = new ThreadLocal<>();
+    
+    @SuppressWarnings("unchecked")
+    private static <T extends Throwable> void sneakyThrow(Throwable exception) throws T {
+            throw (T) exception;
+    }        
+    
+    /**
+     * Retrieves the contextual perf counters. If none are set, returns global counters.
+     * @return context counters, or global ones.
+     */
+    public static PerfData context() {
+        PerfData d = threadInstance.get();
+        return d == null ? global : d;
+    }
+
+    /**
+     * Helper wrapper that establishes a thread-local context instance of {@link PerfData} that can be retrieved by
+     * {@link #context}.
+     * @param <T> result type of the task
+     * @param data perfdata instance
+     * @param c the task to execute
+     * @return data computed by the task
+     */
+    public static <T> T withPerfData(PerfData data, Callable<T> c) {
+        PerfData old = threadInstance.get();
+        try {
+            threadInstance.set(data);
+            return c.call();
+        } catch (Exception ex) {
+            sneakyThrow(ex);
+        } finally {
+            if (old != null) {
+                threadInstance.set(old);
+            } else {
+                threadInstance.remove();
+            }
+        }
+        return null;
+    }
+
+    public PerfData() {
+        collector = global;
+    }
+
+    public void setCollector(PerfData collector) {
+        this.collector = collector;
+    }
+    
+    public void clear() {
+        parserPhaseTime.clear();
+        perfCounters.clear();
+        visitorCounters.clear();
+        fileCount = 0;
+    }
+    
+    public void addParserPhase(int phaseId, long time) {
+        parserPhaseTime.merge(phaseId, time, 
+                (a, b) -> a + b
+        );
+    }
+    
+    public void addVisitorTime(int stage, String visitorKey, long time) {
+        visitorCounters.computeIfAbsent(stage, (s) -> new HashMap<>()).
+            merge(visitorKey, time, (a, b) -> a + b);
+    }
+    
+    public void addPerfCounter(String id, long time) {
+        perfCounters.merge(id, time, (a, b) -> a + b);
+    }
+    
+    public void addPerfStats(PerfData other) {
+        fileCount++;
+        merge(other);
+    }
+    
+    public void merge(PerfData other) {
+        for (Integer i : other.parserPhaseTime.keySet()) {
+            addParserPhase(i, other.parserPhaseTime.get(i));
+        }
+        for (String s : other.perfCounters.keySet()) {
+            addPerfCounter(s, other.perfCounters.get(s));
+        }
+        for (int i = Phases.INITIALIZATION; i <= Phases.ALL; i++) {
+            Map<String, Long> vc = other.visitorCounters.get(i);
+            if (vc == null) {
+                continue;
+            }
+            for (String c : vc.keySet()) { 
+                addVisitorTime(i, c, vc.get(c));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.